source/adapters/cuda/memory.cpp

```diff
@@ -573,6 +419,173 @@ UR_APIEXPORT ur_result_t UR_APICALL urMemBufferPartition(
+}
+
+ur_result_t allocateMemObjOnDeviceIfNeeded(ur_mem_handle_t Mem,
+                                           const ur_device_handle_t hDevice) {
+  ScopedContext Active(hDevice);
+  ur_lock LockGuard(Mem->MemoryAllocationMutex);
+
+  if (Mem->isBuffer()) {
+    auto &Buffer = std::get<BufferMem>(Mem->Mem);
+    auto &DevPtr = Buffer.Ptrs[hDevice->getIndex()];
+
+    // Allocation has already been made
+    if (DevPtr != BufferMem::native_type{0}) {
+      return UR_RESULT_SUCCESS;
+    }
+
+    if (Buffer.MemAllocMode == BufferMem::AllocMode::AllocHostPtr) {
+      // Host allocation has already been made
+      UR_CHECK_ERROR(cuMemHostGetDevicePointer(&DevPtr, Buffer.HostPtr, 0));
+    } else if (Buffer.MemAllocMode == BufferMem::AllocMode::UseHostPtr) {
+      UR_CHECK_ERROR(cuMemHostRegister(Buffer.HostPtr, Buffer.Size,
+                                       CU_MEMHOSTALLOC_DEVICEMAP));
+      UR_CHECK_ERROR(cuMemHostGetDevicePointer(&DevPtr, Buffer.HostPtr, 0));
+    } else {
+      UR_CHECK_ERROR(cuMemAlloc(&DevPtr, Buffer.Size));
+    }
+  } else {
+    CUarray ImageArray;
+    CUsurfObject Surface;
+    try {
+      auto &Image = std::get<SurfaceMem>(Mem->Mem);
+      // Allocation has already been made
+      if (Image.Arrays[hDevice->getIndex()]) {
+        return UR_RESULT_SUCCESS;
+      }
+      UR_CHECK_ERROR(cuArray3DCreate(&ImageArray, &Image.ArrayDesc));
+      Image.Arrays[hDevice->getIndex()] = ImageArray;
+      // HIP_RESOURCE_DESC is a union of different structs, shown here
+      // We need to fill it as described here to use it for a surface or texture
+      // HIP_RESOURCE_DESC::resType must be HIP_RESOURCE_TYPE_ARRAY and
+      // HIP_RESOURCE_DESC::res::array::hArray must be set to a valid HIP array
+      // handle.
+      // HIP_RESOURCE_DESC::flags must be set to zero
+      CUDA_RESOURCE_DESC ImageResDesc;
+      ImageResDesc.res.array.hArray = ImageArray;
+      ImageResDesc.resType = CU_RESOURCE_TYPE_ARRAY;
+
+      UR_CHECK_ERROR(cuSurfObjectCreate(&Surface, &ImageResDesc));
+      Image.SurfObjs[hDevice->getIndex()] = Surface;
+    } catch (ur_result_t Err) {
+      if (ImageArray) {
+        UR_CHECK_ERROR(cuArrayDestroy(ImageArray));
+      }
+      return Err;
+    } catch (...) {
+      if (ImageArray) {
+        UR_CHECK_ERROR(cuArrayDestroy(ImageArray));
+      }
+      return UR_RESULT_ERROR_UNKNOWN;
+    }
+  }
+  return UR_RESULT_SUCCESS;
+}
+
+namespace {
+ur_result_t migrateBufferToDevice(ur_mem_handle_t Mem,
+                                  ur_device_handle_t hDevice) {
+  auto &Buffer = std::get<BufferMem>(Mem->Mem);
+  if (Mem->LastEventWritingToMemObj == nullptr) {
+    // Device allocation being initialized from host for the first time
+    if (Buffer.HostPtr) {
+      UR_CHECK_ERROR(
+          cuMemcpyHtoD(Buffer.getPtr(hDevice), Buffer.HostPtr, Buffer.Size));
+    }
+  } else if (Mem->LastDeviceWritingToMemObj != hDevice) {
+    UR_CHECK_ERROR(cuMemcpyDtoD(Buffer.getPtr(hDevice),
```

Comment:
nit: cuMemcpyDtoD does not copy directly between devices, the semantics of it is device A to device A. I vaguely remember that it does work to copy device A to device B via host, but that if i remember correctly it was slower than manually doing cuMemcpyDtoH for device A to host then cuMemcpyHtoD for host to device B.

But once you get into this then it raises the issue that if you want best performance copies should be done P2P, which raises a whole host of other issues like max peers etc. So probably this isn't something worth worrying about at this point.

Reply:
I think I'm happy with leaving this as is. If there is some perf gain to using DtoD vs DtoH + HtoD I can make a ticket to benchmark this difference.

Reply:
One other way would be to put two cuda devices into different "sycl contexts": the sycl runtime is still responsible for doing the copy for such a case, and it does it via D2H H2D. So then if that is not faster than doing it for two devs within a single "sycl context" then probs we are all good.
I think doing this performance comparison (two devices in different sycl contexts and two devices within one) would be good irrespective, just to confirm that there is no other potential performance regression issue beyond the the D2H H2D vs "D2D" one I flagged.